service/raft/raft_group_registry.cc

```diff
@@ -24,22 +25,9 @@ raft_group_registry::raft_group_registry(bool is_enabled, netw::messaging_servic
     : _is_enabled(is_enabled)
     , _ms(ms)
     , _fd(make_shared<raft_gossip_failure_detector>(gossiper, _srv_address_mappings))
-    , _raft_support_listener(feat.cluster_supports_raft_cluster_mgmt().when_enabled([this, &feat, &gossiper] {
-        // If the `USES_RAFT_CLUSTER_MANAGEMENT` feature was already enabled, do nothing.
-        //
-        // This can happen if the current node is either a fresh node in an empty cluster
-        // or a restarting node, which is already part of an existing group0.
-        if (!_is_enabled || feat.cluster_uses_raft_cluster_mgmt()) {
-            return;
-        }
-        // When the cluster fully supports raft-based cluster management,
-        // we can re-enable support for the second gossip feature to trigger
-        // actual use of raft-based cluster management procedures.
-        feat.support(gms::features::USES_RAFT_CLUSTER_MANAGEMENT).get();
-        if (this_shard_id() == 0) {
-            // When the supported feature set is dynamically extended, re-advertise it through the gossip.
-            gossiper.add_local_application_state(gms::application_state::SUPPORTED_FEATURES,
-                gms::versioned_value::supported_features(feat.supported_feature_set())).get();
+    , _raft_support_listener(feat.cluster_supports_raft_cluster_mgmt().when_enabled([this] {
+        if (_group0) {
+            _group0->join_group0().get();
```

Comment:
- join_group0() currently uses the list of seeds (from scylla.yaml) to initialize discovery. But in upgrade we must ensure that **every** member of the Scylla cluster is included in the newly formed Raft cluster. We cannot create group 0 with some arbitrary subset of nodes, some of which may no longer exist (the seed list is only used for joining the cluster) and which may lead to separate group 0 clusters being formed (different nodes may use different disjoint seed lists).
- We must not block gossiper on join group 0. The listener is called from gossiper (`do_on_change_notifications`); right now the code will block gossiper (this is what caused deadlocks earlier) join_group0() may be a long operation, it needs to run the distributed discovery algorithm, all nodes must contact each other (see also first bullet).